--- origsrc/libgtop-2.24.0/examples/netload.c	2008-05-23 17:13:26.000000000 -0500
+++ src/libgtop-2.24.0/examples/netload.c	2008-12-15 15:08:16.259500000 -0600
@@ -67,7 +67,9 @@
 	unsigned method, count, port;
 	struct in_addr addr, subnet;
 	char *address_string, *subnet_string;
+#ifdef AF_INET6
 	char address6_string[INET6_ADDRSTRLEN], prefix6_string[INET6_ADDRSTRLEN];
+#endif
 	char *hwaddress_string;
 	char buffer [BUFSIZ];
 
@@ -108,8 +110,10 @@
 	address_string = g_strdup (inet_ntoa (addr));
 	subnet_string  = g_strdup (inet_ntoa (subnet));
 
+#ifdef AF_INET6
 	inet_ntop (AF_INET6, netload.address6, address6_string, INET6_ADDRSTRLEN);
 	inet_ntop (AF_INET6, netload.prefix6,  prefix6_string,  INET6_ADDRSTRLEN);
+#endif
 
 	hwaddress_string = hwaddress_format_for_display(&netload);
 
@@ -128,9 +132,11 @@
 		"\tErrors In:\t\t%llu\n"
 		"\tErrors Out:\t\t%llu\n"
 		"\tErrors Total:\t\t%llu\n\n"
+#ifdef AF_INET6
 		"\tAddress6:\t\t%s\n"
 		"\tPrefix6:\t\t%s\n"
 		"\tScope6:\t\t\t%#03x\n\n"
+#endif
 		"\tHarware Address:\t%s\n\n",
 		netload.flags,
 		netload.if_flags,
@@ -147,9 +153,11 @@
 		netload.errors_in,
 		netload.errors_out,
 		netload.errors_total,
+#ifdef AF_INET6
 		address6_string,
 		prefix6_string,
 		(int) netload.scope6,
+#endif
 		hwaddress_string);
 
 
