--- origsrc/libgtop-2.24.0/lib/Makefile.am	2008-05-23 17:13:26.000000000 -0500
+++ src/libgtop-2.24.0/lib/Makefile.am	2008-12-15 13:44:50.556375000 -0600
@@ -7,7 +7,7 @@
 				  read_data.c write.c lib.c parameter.c \
 				  sysdeps.c
 
-libgtop_2_0_la_LDFLAGS		= $(LT_VERSION_INFO) -export-symbols $(srcdir)/libgtop.sym
+libgtop_2_0_la_LDFLAGS		= $(LT_VERSION_INFO) -export-symbols $(srcdir)/libgtop.sym -no-undefined
 libgtop_2_0_la_LIBADD		= $(GLIB_LIBS) $(top_builddir)/sysdeps/common/libgtop_common-2.0.la $(top_builddir)/sysdeps/@sysdeps_dir@/libgtop_sysdeps-2.0.la $(sysdeps_suid_lib)
 
 BUILT_SOURCES			= lib.c
